pinot-core/src/main/java/com/linkedin/pinot/core/common/DataFetcher.java

```diff
@@ -0,0 +1,190 @@
+/**
+ * Copyright (C) 2014-2015 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.core.common;
+
+import com.linkedin.pinot.core.indexsegment.IndexSegment;
+import com.linkedin.pinot.core.plan.DocIdSetPlanNode;
+import com.linkedin.pinot.core.segment.index.readers.Dictionary;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+
+/**
+ * DataFetcher is a higher level abstraction for data fetching. Given an index segment, DataFetcher can manage the
+ * DataSource, Dictionary and BlockValSet for this segment, preventing redundant construction for these instances.
+ * Initialize DataFetcher with block doc id set and block length, DataFetcher will take care of the data fetching
+ * process and reuse resources if possible. DataFetcher can be used by both selection, aggregation and group-by data
+ * fetching process, reducing duplicate codes and garbage collection.
+ *
+ * TODO: Support Multi Value Columns
+ */
+public class DataFetcher {
```

Comment:
have a interface and two impls for singlevalueDataFetcher and one for multivalue since most of the apis will be mutually exclusive